src/mlpack/core/tree/vantage_point_tree/vantage_point_split_impl.hpp

```diff
@@ -0,0 +1,298 @@
+    // decrease the right counter accordingly.  When we encounter one that isn't
+    // correct, stop.  We will switch it with the wrong point we found in the
+    // previous loop.
+    while ((!AssignToLeftSubtree(bound, data, vantagePoint, right, mu)) && (left <= right))
+      right--;
+  }
+
+  Log::Assert(left == right + 1);
+
+  return left;
+}
+
+template<typename BoundType, typename MatType>
+template<typename VecType>
+size_t VantagePointSplit<BoundType, MatType>::PerformSplit(const BoundType& bound,
+    MatType& data,
+    const size_t begin,
+    const size_t count,
+    const VecType& vantagePoint,
+    const ElemType mu,
+    std::vector<size_t>& oldFromNew)
+{
+  // This method modifies the input dataset.  We loop both from the left and
+  // right sides of the points contained in this node.  The points closer to
+  // the vantage point should be on the left side of the matrix, and the farther
+  // from the vantage point should be on the right side of the matrix.
+  size_t left = begin;
+  size_t right = begin + count - 1;
+
+  // First half-iteration of the loop is out here because the termination
+  // condition is in the middle.
+
+  while (AssignToLeftSubtree(bound, data, vantagePoint, left, mu) && (left <= right))
+    left++;
+
+  while ((!AssignToLeftSubtree(bound, data, vantagePoint, right, mu)) && (left <= right) && (right > 0))
+    right--;
+
+  while (left <= right)
+  {
+    // Swap columns.
+    data.swap_cols(left, right);
+
+    // Update the indices for what we changed.
+    size_t t = oldFromNew[left];
+    oldFromNew[left] = oldFromNew[right];
+    oldFromNew[right] = t;
+
+    // See how many points on the left are correct.  When they are correct,
+    // increase the left counter accordingly.  When we encounter one that isn't
+    // correct, stop.  We will switch it later.
+    while (AssignToLeftSubtree(bound, data, vantagePoint, left, mu) && (left <= right))
+      left++;
+
+    // Now see how many points on the right are correct.  When they are correct,
+    // decrease the right counter accordingly.  When we encounter one that isn't
+    // correct, stop.  We will switch it with the wrong point we found in the
+    // previous loop.
+    while ((!AssignToLeftSubtree(bound, data, vantagePoint, right, mu)) && (left <= right))
+      right--;
+  }
+
+  Log::Assert(left == right + 1);
+
+  return left;
+}
+
+template<typename BoundType, typename MatType>
+void VantagePointSplit<BoundType, MatType>::
+SelectVantagePoint(const BoundType& bound, const MatType& data,
+    const size_t begin, const size_t count, size_t& vantagePoint, ElemType& mu)
+{
+  arma::uvec vantagePointCandidates;
+
+  // Get no more than max(maxNumSamples, count) vantage point candidates
+  GetDistinctSamples(vantagePointCandidates, maxNumSamples, begin, count);
+
+  ElemType bestSpread = 0;
+
+  //  Evaluate eache candidate
+  for (size_t i = 0; i < vantagePointCandidates.n_rows; i++)
+  {
+    arma::uvec samples;
```

Comment:
I guess it is better to define all vectors as private members of the class. It is possible since we use an instantiated object. But we have to make all methods non-static. How do you think?